core/src/commonMain/kotlin/com/powersync/db/internal/ConnectionPool.kt

```diff
@@ -0,0 +1,152 @@
+package com.powersync.db.internal
+
+import co.touchlab.stately.concurrency.AtomicBoolean
+import com.powersync.PowerSyncException
+import com.powersync.PsSqlDriver
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.joinAll
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+internal class WrappedConnection(
+    val transactorDriver: TransactorDriver,
+) {
+    var busy: Boolean = false
+}
+
+internal data class DeferredAction<R>(
+    val deferred: CompletableDeferred<R>,
+    val action: suspend (connection: TransactorDriver) -> R,
+)
+
+internal class ConnectionPool(
+    factory: () -> PsSqlDriver,
+    size: Int = 5,
+    private val scope: CoroutineScope,
+) {
+    val closed = AtomicBoolean(false)
+
+    private val mutex = Mutex()
+    private val connections = List(size) { WrappedConnection(TransactorDriver(factory())) }
+    private val queue = mutableListOf<DeferredAction<Any?>>()
+    private val activeOperations = mutableListOf<CompletableDeferred<Any?>>()
+
+    @Suppress("UNCHECKED_CAST")
+    suspend fun <R> withConnection(action: suspend (connection: TransactorDriver) -> R): R {
+        if (closed.value) {
+            throw PowerSyncException(
+                message = "Cannot process connection pool request",
+                cause = Exception("Pool is closed"),
+            )
+        }
+
+        val wrappedDeferred = DeferredAction(CompletableDeferred<Any?>(), action)
```

Comment:
I think this might have the downside that `action` might be executed in a different coroutine scope, breaking cancellations? I'm not sure about this though.

Reply:
Yeah, cancellations would have been an issue. I've updated the pooling implementation to the recommended channel approach which should not have this issue.